### DISCUSSION TITLE
Keep workflow trigger after auto-commit on external contribution PR

### DISCUSSION THREAD

**Discussion by WarningImHack3r** [stefanzweifel/git-auto-commit-action#292] [Help] [answered]
I have a workflow that checks out the repo (with a `pull_request` trigger), makes some pre-checks and then execute the build step etc.

My issue is the following: for a PR I make on my repo from another branch, everything works fine and I have the intended behavior.   
However, when the workflow is triggered from an external member (dependabot or forks), the workflow cannot start because I use a `secrets.PAT` [as recommended in the README](https://github.com/stefanzweifel/git-auto-commit-action#commits-made-by-this-action-do-not-trigger-new-workflow-runs) to allow the workflow to be triggered from the new commits. This secret cannot be fetched from there for obvious security reasons on an unknown codebase.

A workaround could be to use the `pull_request_target` trigger, however in that case I can indeed access my secrets but the checkout would occur on _my_ `main` branch, which is not what I want.

To sum up, is there a way to keep a workflow triggered from the auto-commit without exposing secrets to an external contribution?

**Comment by stefanzweifel** [comment, marked as answer]
Hey

We have a dedicated section about how to setup the Action to run for forks of public repositories here: https://github.com/stefanzweifel/git-auto-commit-action#use-in-forks-from-public-repositories

In that workflow, there is no need for a PAT, if you don't use protected branches.
Could you give that a try?

Personal access token are ususally only needed if your repository uses protected branches or you must trigger another workflow run from commits made with this action.

If you must have secrets in your workflow and they need to be run for contributors, I actually don't know how to do that either. Would also need to do some research on this.

(In a personal project of mine, I also must [use secrets for an AWS account](https://github.com/stefanzweifel/sidecar-browsershot/blob/a758f2419931f8bb145a9754774fc0bc3dd9baba/.github/workflows/run-tests.yml#L68-L72) and I haven't figured out yet, how do make them safely available to contributors or dependabot. There is a special "Dependabot" section for secrets, but I don't think that works for contributors using public forks)

**Reply by WarningImHack3r** [reply]
Thanks for that! I didn't see it while reading the README, sorry.  
So actually what matters most is to check out the PR's repo, so the [first solution](https://github.com/stefanzweifel/git-auto-commit-action#workflow-should-run-in-base-repository) should do the trick just fine!

FYI:
- My `main` branch is indeed protected (not to allow direct push) but it shouldn't change anything I guess
- I don't use any other `secrets.*`

So yeah, I will try this first solution and update here if something doesn't work well. Thanks!

**Reply by WarningImHack3r** [reply]
Update @stefanzweifel: this setup will indeed work in forks because it doesn't use secrets anymore, but now the workflow isn't triggered anymore by the automatically pushed commit... What's your guess?
```yaml
jobs:
  myjob:
    runs-on: ubuntu-latest
    permissions:
      contents: write
    
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          repository: ${{ github.event.pull_request.head.repo.full_name }}
          ref: ${{ github.head_ref }}
      
      ...

      - name: Commit
        uses: stefanzweifel/git-auto-commit-action@v4
        with:
          commit_message: Update file
```

**Reply by stefanzweifel** [reply]
@WarningImHack3r I think now we're back at the limitation by GitHub, that commits that are created with the default `github_token` will never trigger another workflow run. 😐 

I really don't know what the best approach here would be. You would need to use a PAT to push the commit, but that is stored as a secret, which you don't want to / can't expose.

**Reply by WarningImHack3r** [reply]
Ok so I have two options:
1) Use `pull_request_target` but I'm not sure if it's going to break the workflow behavior or not, would you know? Would it even solve the problem?
2) Put the "fix steps" under a condition (which would be "if not a fork and not dependabot" I guess) and do not fix stuff for these cases and simply throw an error instead

**Reply by stefanzweifel** [reply]
I just read through some old discussions around this issue. In #211 we had a very long and detailed conversation about this.

In a demo repo of mine, my 2nd GitHub account created [a PR](https://github.com/stefanzweifel/git-auto-commit-action-demo-app/pull/34) and correctly triggered git-auto-commit and pushed a commit to GitHub. 
[The workflow](https://github.com/stefanzweifel/git-auto-commit-action-demo-app/blob/11eabf3ff0c22154c221ae8f0ad4a84075f3d3bd/.github/workflows/format_php.yml) is listening to the `pull_request_target` event and also checks out the head repo. There is no need for a personal access token.

The PAT is really only necessary, if you need to push to a protected branch **or** the generated commit must/should trigger other workflow runs. (In your case, maybe running a test suite after the software has been built and pushed back to GitHub.)

Regarding secrets: I think secret values are never ever shown in logs or can be read in a way to display them. (But I'm not 100% sure here. Would have to re-read the docs for this feature)

---

Selecting the next option for you depends on what you want to achieve with your workflow. 
I would go with option 2.

I've created similar workflows in the past, that only run if the PR has been created by Dependabot. Here's an example from a reusable workflow: https://github.com/2media/reusable-actions-workflows/blob/main/.github/workflows/integrate-landingpages-frontend.yml#L18

The condition can easily be negated by using `if: github.actor != 'dependabot[bot]'`

---

I know this isn't the best answer. GitHub Actions + git-auto-commit + Forks is quite a complicated scenario and GitHub doesn't make it easy to understand. 😐

**Reply by WarningImHack3r** [reply]
For option 1 and the PR you linked, it maybe works but it does not trigger the workflow again from the new auto-commit, which is basically my current situation and what I want to avoid:
- A commit without workflow blocks the ability to merge the PR if you have a required workflow set up on your repo (AFAIK)
- It does not guarantee that the auto-commit fixes everything and builds successfully, a new workflow run is mandatory

So, as the "for everybody" + auto-commit + "auto-commit triggers workflow" setup does not seem possible, I'm going for option 2 unless a solution is discovered someday or GitHub addresses this feature lack.

---

For option 2, putting an `if:` statement on top of the job is indeed straightforward. However, as the target setup would be the following...:
- If _allowed_, checkout with PAT & `token:` and push the auto-commit which will trigger the workflow once again, all good
- If _not allowed_, checkout with `ref:` and `repository:`, and simply throw an error if some file is not correct, letting the user fix it manually (painful for Dependabot with lock files, but no other choice)

...what are the conditions that prevent a case scenario from being _allowed_? It's confusing because Dependabot creates a new branch on the original repo without needing any fork but still doesn't have permission to fetch `secrets` (and to be fair this whole problem is very confusing)...

At the end of the day, wouldn't the condition be `if the actor is not a member of the organization/contributors`? If it's the right condition, how can I write it? (A bit out of your scope I know but you're the most likely person to be able to know that)

**Reply by stefanzweifel** [reply]
I agree: it's all very confusing. Especially the differentiation between "normal" users and Dependabot.
(In past projects, I had to duplicate my secrets in the repository setting, as there are "normal secrets" and "Dependabot secrets".)

The idea of detecting if the user is in the organisation or is a contributor makes a lot of sense.
Unfortunately the [`github`-context](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context) doesn't return that much user information. Only the username and user_id is returned. With additional API requests one could probably figure out, if the user has the right permissions.

**Reply by WarningImHack3r** [reply]
I just came back to that problem to try implementing it. However, I'm still stuck on that _if allowed_ condition because I want the best implementation and fewer drawbacks. Did you try getting in touch with the GH team to get their advice/explanations? Should I give it a(nother) try for this specific problem I'm trying to solve?

**Reply by stefanzweifel** [reply]
> Did you try getting in touch with the GH team to get their advice/explanations?

I don't have direct access to anyone of GitHub. The best place for such problems/questions is probably the community discussion board: https://github.com/orgs/community/discussions/categories/actions

> Should I give it a(nother) try for this specific problem I'm trying to solve?

I can't stop you. :)
Personally, I think this is quite a hard problem to solve, as GitHub as a platform makes it hard. But if you would solve this hard problem and publish your findings/solutions somewhere, you would make many GH Actions users happy.

**Reply by WarningImHack3r** [reply]
Hi @stefanzweifel! I managed to build up a "kinda clean" solution [here](https://github.com/EmeraldHQ/Website/blob/97d09d81d54439fdee42ef1221f30b95b8a3d085/.github/workflows/pages-deploy.yml#L28-L48). The idea is simple: try to check out a first time and [adapt the rest of the jobs](https://github.com/EmeraldHQ/Website/blob/97d09d81d54439fdee42ef1221f30b95b8a3d085/.github/workflows/pages-deploy.yml#L53-L54) and/or [the steps](https://github.com/EmeraldHQ/Website/blob/97d09d81d54439fdee42ef1221f30b95b8a3d085/.github/workflows/pages-deploy.yml#L64) depending on if it fails or not! This is the "_if allowed_" condition I mentioned above.
Also, note that I added the same exact `secrets.WORKFLOW_PAT` to both the org secrets and org dependabot's secrets.

Overall, it does the job as intended but it's not the cleanest solution on earth so I'm only half-proud of it. Anyway, thanks for your help!
